src/MSIdentityScaffolding/Microsoft.DotNet.MSIdentity/CodeReaderWriter/ProjectModifier.cs

```diff
@@ -157,107 +166,68 @@ private void AddFile(CodeFile file)
         {
-            var document = project.Documents.Where(d => d.Name.EndsWith(fileName)).FirstOrDefault();
-            if (document is null)
+            if (file.FileName.Equals("Startup.cs"))
+            {
+                // Startup class file name may be different
+                file.FileName = await ProjectModifierHelper.GetStartupClass(project) ?? file.FileName;
+            }
+
+            var fileDoc = project.Documents.Where(d => d.Name.Equals(file.FileName)).FirstOrDefault();
+            if (fileDoc is null)
             {
                 return;
             }
 
-            var razorChanges = file?.RazorChanges.Where(cc => ProjectModifierHelper.FilterOptions(cc.Options, toolOptions));
-            var editedDocument = await ProjectModifierHelper.ModifyDocumentText(document, razorChanges);
-            await ProjectModifierHelper.UpdateDocument(editedDocument, _consoleLogger);
-        }
+            //get the file document to get the document root for editing.
+            DocumentEditor documentEditor = await DocumentEditor.CreateAsync(fileDoc);
+            if (documentEditor is null)
+            {
+                return;
+            }
 
-        internal async Task ModifyCshtmlFile(string fileName, CodeFile cshtmlFile, CodeAnalysis.Project project, CodeChangeOptions options)
-        {
-            string? filePath = Directory.EnumerateFiles(_toolOptions.ProjectPath, fileName, SearchOption.AllDirectories).FirstOrDefault();
+            DocumentBuilder documentBuilder = new DocumentBuilder(documentEditor, file, _consoleLogger);
+            var modifiedRoot = ModifyRoot(documentBuilder, options, file);
 
-            if (!string.IsNullOrEmpty(filePath))
+            if (modifiedRoot != null)
             {
-                var fileDoc = project.Documents.Where(d => d.Name.Equals(filePath)).FirstOrDefault();
-                if (fileDoc != null)
-                {
-                    //add code snippets/changes.
-                    if (cshtmlFile.Methods != null && cshtmlFile.Methods.Any())
-                    {
-                        var globalChanges = cshtmlFile.Methods.TryGetValue("Global", out var globalMethod);
-                        if (globalMethod != null)
-                        {
-                            var filteredCodeFiles = globalMethod.CodeChanges.Where(cc => ProjectModifierHelper.FilterOptions(cc.Options, options));
-                            var editedDocument = await ProjectModifierHelper.ModifyDocumentText(fileDoc, filteredCodeFiles);
-                            //replace the document
-                            await ProjectModifierHelper.UpdateDocument(editedDocument, _consoleLogger);
-                        }
-                    }
-                }
+                documentEditor.ReplaceNode(documentEditor.OriginalRoot, modifiedRoot);
+            }
+
+            if (!string.IsNullOrEmpty(fileDoc.FilePath))
```

Comment:
Let's move this check up to line 178. if `fileDoc.FilePath` is null/empty, let's not do all this extra work.